sw/device/lib/dif/dif_edn.c

```diff
@@ -68,27 +68,75 @@ dif_result_t dif_edn_set_auto_mode(const dif_edn_t *edn,
+  uint32_t ctrl_reg = mmio_region_read32(edn->base_addr, EDN_CTRL_REG_OFFSET);
+  const uint32_t edn_en =
+      bitfield_field32_read(ctrl_reg, EDN_CTRL_EDN_ENABLE_FIELD);
+  if ((enum multi_bit_bool)edn_en != kMultiBitBool4False) {
+    return kDifError;
+  }
 
+  // Ensure neither automatic nor boot request mode is set.
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_AUTO_REQ_MODE_FIELD,
+                                    kMultiBitBool4False);
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_BOOT_REQ_MODE_FIELD,
+                                    kMultiBitBool4False);
+  mmio_region_write32(edn->base_addr, EDN_CTRL_REG_OFFSET, ctrl_reg);
+
+  // Clear the reseed command FIFO and the generate command FIFO.
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_CMD_FIFO_RST_FIELD,
+                                    kMultiBitBool4True);
+  mmio_region_write32(edn->base_addr, EDN_CTRL_REG_OFFSET, ctrl_reg);
+
+  // Disable clearing of those FIFOs.  This is a prerequisite before any
+  // further commands can be issued to these FIFOs.
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_CMD_FIFO_RST_FIELD,
+                                    kMultiBitBool4False);
+  mmio_region_write32(edn->base_addr, EDN_CTRL_REG_OFFSET, ctrl_reg);
+
+  // Fill the reseed command FIFO.
   for (size_t i = 0; i < config.reseed_material.len; ++i) {
     mmio_region_write32(edn->base_addr, EDN_RESEED_CMD_REG_OFFSET,
                         config.reseed_material.data[i]);
   }
+
+  // Fill the generate command FIFO.
   for (size_t i = 0; i < config.generate_material.len; ++i) {
     mmio_region_write32(edn->base_addr, EDN_GENERATE_CMD_REG_OFFSET,
                         config.generate_material.data[i]);
   }
 
+  // Set the maximum number of requests between reseeds.
   mmio_region_write32(edn->base_addr,
                       EDN_MAX_NUM_REQS_BETWEEN_RESEEDS_REG_OFFSET,
                       config.reseed_interval);
 
-  reg = bitfield_field32_write(reg, EDN_CTRL_AUTO_REQ_MODE_FIELD,
-                               kMultiBitBool4True);
-  mmio_region_write32(edn->base_addr, EDN_CTRL_REG_OFFSET, reg);
+  // Re-enable EDN in automatic request mode.
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_EDN_ENABLE_FIELD,
+                                    kMultiBitBool4True);
+  ctrl_reg = bitfield_field32_write(ctrl_reg, EDN_CTRL_AUTO_REQ_MODE_FIELD,
+                                    kMultiBitBool4True);
+  mmio_region_write32(edn->base_addr, EDN_CTRL_REG_OFFSET, ctrl_reg);
+
+  // Wait until EDN is ready to accept commands.
+  while (true) {
+    uint32_t reg =
+        mmio_region_read32(edn->base_addr, EDN_SW_CMD_STS_REG_OFFSET);
+    if (bitfield_bit32_read(reg, EDN_SW_CMD_STS_CMD_RDY_BIT)) {
+      break;
+    }
+  }
+
+  // Command CSRNG Instantiate.  As soon as CSRNG acknowledges this command,
+  // EDN will start automatically sending reseed and generate commands.
+  DIF_RETURN_IF_ERROR(csrng_send_app_cmd(
```

Comment:
do we think this call should be inside this particular dif function? There is another `dif_instantiate` call, i wonder if we should just leave it to the wrapping function / test to invoke.  

That way it can poll for done also and we won't have to bring the polling function into the dif

Reply:
Assuming there will be no change in the code due to the comments above and after this function call we just return, then the macro `DIF_RETURN_IF_ERROR` should be removed:
```c
return csrng_send_app_cmd(
...
```

Reply:
> do we think this call should be inside this particular dif function?

Thanks for pointing this out, @tjaychen.

**Current variant**: I included the instantiation command (which I agree should go through `dif_edn_instantiate()`) inside `dif_edn_set_auto_mode()` to make it as self-contained as `dif_edn_set_boot_mode()`. To complete this, I would add polling with `dif_edn_get_status(.., EDN_SW_CMD_STS_CMD_STS_BIT, ..)` at the end of `dif_edn_set_auto_mode()`. This way, the caller can be sure that the EDN operates in auto mode when the function returns.

**Alternative variant**: We could require users to call `dif_edn_instantiate()` and poll `dif_edn_get_status()` after `dif_edn_set_auto_mode()`. I think this is further away from the DIF philosophy "make it easy to use HW as intended and difficult to misuse it", though, so I'd prefer the current variant.

If we decide on the current variant, we might want to discuss if `dif_edn_set_auto_mode()` should get additional arguments to let users influence the seed of the instantiate command (which could be as simple as making `entropy_src_enable` and `seed_material` arguments).

Reply:
I vote for the current variant, and I like the idea of adding additional arguments (but optional) to the dif.

Reply:
sgtm!